From edcd4c2bd7886d5148f302a53e1133b9dfb425a6 Mon Sep 17 00:00:00 2001
From: Harald Hoyer <harald@redhat.com>
Date: Fri, 5 Aug 2011 13:05:21 +0200
Subject: [PATCH] 90lvm/lvm_scan.sh: use "--partial" to force assembly
 incomplete VGs

If our internal loop counter is bigger than half of the maximum
count, try to assemble lvm device partially with the "--partial"
option.

https://bugzilla.redhat.com/show_bug.cgi?id=723548
---
 modules.d/90lvm/64-lvm.rules | 1 +
 modules.d/90lvm/lvm_scan.sh  | 8 ++++++--
 2 files changed, 7 insertions(+), 2 deletions(-)

diff --git a/modules.d/90lvm/64-lvm.rules b/modules.d/90lvm/64-lvm.rules
index 1c28b54..3bd46f4 100644
--- a/modules.d/90lvm/64-lvm.rules
+++ b/modules.d/90lvm/64-lvm.rules
@@ -14,6 +14,7 @@ PROGRAM=="/bin/sh -c 'for i in $sys/$devpath/holders/dm-[0-9]*; do [ -e $$i ] &&
     GOTO="lvm_end"
 
 RUN+="/sbin/initqueue --settled --onetime --unique /sbin/lvm_scan"
+RUN+="/sbin/initqueue --timeout --onetime --unique /sbin/lvm_scan --partial"
 RUN+="/bin/sh -c '>/tmp/.lvm_scan-%k;'"
 
 LABEL="lvm_end"
diff --git a/modules.d/90lvm/lvm_scan.sh b/modules.d/90lvm/lvm_scan.sh
index 9f17d48..0938041 100755
--- a/modules.d/90lvm/lvm_scan.sh
+++ b/modules.d/90lvm/lvm_scan.sh
@@ -2,6 +2,7 @@
 
 # run lvm scan if udev has settled
 
+extraargs="$@"
 . /lib/dracut-lib.sh
 
 VGS=$(getargs rd_LVM_VG=)
@@ -60,8 +61,11 @@ sub=${sub%%\(*};
 check_lvm_ver 2 2 57 $maj $min $sub && \
     nopoll="--poll n"
 
-check_lvm_ver 2 2 65 $maj $min $sub && \
-    sysinit=" --sysinit "
+if check_lvm_ver 2 2 65 $maj $min $sub; then
+    sysinit=" --sysinit $extraargs"
+fi
+
+unset extraargs
 
 export LVM_SUPPRESS_LOCKING_FAILURE_MESSAGES=1
 
